lms/security.py

```diff
@@ -197,7 +207,34 @@ def _get_user(request):
     )
 
 
+def _get_lti_jwt(request):
+    id_token = request.params.get("id_token")
+    if not id_token:
+        return {}
+
+    try:
+        jwt_params = jwt.decode(id_token, options={"verify_signature": False})
+    except InvalidJWTError as err:
+        raise marshmallow.ValidationError("Invalid id_token", "authorization") from err
+
+    warnings.warn("Using not verified JWT token")
+    return jwt_params
+
+
+def _get_lti_params(request) -> LTIParams:
+    """Return the requests LTI parameters."""
+    return (
+        LTIParams.from_v13(request.lti_jwt)
+        if request.lti_jwt
+        else LTIParams(request.params)
+    )
+
+
 def includeme(config):
     config.set_security_policy(SecurityPolicy(config.registry.settings["lms_secret"]))
     config.add_request_method(_get_lti_user, name="lti_user", property=True, reify=True)
+    config.add_request_method(_get_lti_jwt, name="lti_jwt", property=True, reify=True)
     config.add_request_method(_get_user, name="user", property=True, reify=True)
+    config.add_request_method(
+        _get_lti_params, name="lti_params", property=True, reify=True
+    )
```

Comment:
I'm pretty skeptical of request methods, as they are very non-local by nature and it's quite difficult to put together how they are used. As they appear on the request object it can be difficult to immediately even notice they aren't part of `pyramid`. In this case they are prefixed with `lti_` which should tip the reader off, which you can then try and grep for the source (which can be anywhere).

In this case `lti_jwt` only seems to be used in the `_get_lti_params` method. So it's defined for the whole app, but is it, practically speaking, a private variable for this function? `lti_params` also seems to only be used in the context object in `lms/resources/lti_launch.py` so seems to be functionally private to that too?

My current understanding is this is defined in `security` and added on all requests, even those which don't have any LTI params, to be used in a context which is only available in one set of views? Should this be in the context object?

I could see how `lti_params` could be more generally useful. Is this the start of moving away from `parsed_params`? Otherwise this seems to tread the exact same ground?

Reply:
I've just spotted that these are used as `self.context['request'].lti_jwt`. I was looking for `request.lti_jwt` for usages.

I'm 100% unsure if this means everything I said is wrong, or kind of proves my point.

Reply:
I think `lti_params` makes sense in  the request, it's  going to be used in views,resource and predicates.

`lti_jwt` not so much. It belongs to a service JWT service with a `@cache` decorator but as I mention somewhere here that's a big TODO, that service it bigger that what is present here and will required IMO moving other JWT usages to it so I left it for another PR if that makes sense.

Reply:
We had a chat about this, and I think we agreed it only makes sense in the same places that the `resources/lti_launch.py` does, rather than all views. This means this probably can be sucked up into the `LTILaunchResource`.